tests/integration/sandbox.py

```diff
@@ -762,6 +770,216 @@ def restart(self):
+                val += str(ops[i])
+            elif ops[i].op_type == "read":
+                tmp = " ".join([x.decode('utf-8') for x in results[i]])
+                val += f"[:r \"{ops[i].key}\" [{tmp}]]"
+            else:
+                raise Exception(f"unknown op type {ops[i].op_type}")
+        val += "]"
+        return val
+
+    def generate_result_output(self, thread: int, result: str, results: typing.Optional[List], ops: List[ElleOp]):
+        output: str
+        if result == "ok":
+            output = self.generate_results_value(results, ops)
+            return f"{{:index {self.index[thread]} :process {thread} :type :{result}, :value {output}}}"
+        elif result == "fail":
+            output = self.generate_ops_value(ops)
+            return f"{{:index {self.index[thread]} :process {thread} :type :{result}, :value {output}}}"
+        else:
+            raise Exception(f"unknown result value {result}")
+
+    def log_result(self, thread, result: str, results: typing.Optional[List], ops: List[ElleOp]):
+        # if this throws an exception, we have a failure elsewhere
+        self.index[thread] += 1
+
+        output = self.generate_result_output(thread, result, results, ops)
+        self.lock.acquire()
+        self.logfile.write(output)
+        self.logfile.write("\n")
+        self.lock.release()
+
+    def log_comment(self, msg: str):
+        self.lock.acquire()
+        self.logfile.write("; " + msg + "\n")
+        self.lock.release()
+
+    @staticmethod
+    def map_addresses_to_clients(clusters: List[Cluster]) -> typing.Dict[str, redis.Redis]:
+        mapped_clients: typing.Dict[str, redis.Redis] = {}
+        default_set = False
+        for c in clusters:
+            for n in c.nodes.values():
+                mapped_clients[n.address] = n.client
+                if not default_set:
+                    mapped_clients["default"] = n.client
+                    default_set = True
+        return mapped_clients
+
+    def get_next_value(self) -> int:
+        with self.lock:
+            ret = self.value
+            self.value += 1
+            return ret
+
+
+class ElleWorker(threading.Thread):
+    def __init__(self, elle: Elle, clusters: typing.List[Cluster], keys: typing.Optional[List[str]] = None):
+        super().__init__()
+
+        self.clusters = clusters
+        self.elle: Elle = elle
+        self.ended: bool = False
+        self.id: int = 0
+        self.keys: typing.List
+        if keys is not None:
+            self.keys = keys
+        else:
+            self.keys = ["test"]
+
+    def finish(self):
+        self.ended = True
+
+    def set_keys(self, keys: typing.List):
+        self.keys = keys
+
+    # might not want generate_ops, perhaps each multi should just be a single append/read of same key
+    def generate_ops(self, available_keys: typing.List[str]) -> typing.List[ElleOp]:
+        ops: typing.List[ElleOp] = []
+        for _ in range(random.randint(1, 1)):
```

Comment:
Probably, related with the above. This is not a loop right? `random.randint(1,1)` will generate `1` only.

Reply:
so, my thought here was that it might be larger, as we might have multi/exec transactions with more than one append/read pair.  I'm not sure if it gains us anything.  If it doesn't, can remove the loop.

Reply:
Ok, it is up to you.

Reply:
I'm going to make it configurable, but will default to 1

Reply:
and no need for random.  random doesn't give us anything here.  so will be configurable to be larger than 1 append/read pair, but will default to 1.